doc/installation/common_configuration.rst
Common configuration options
===================================

Customizing resource record types
--------------------------------------

If you're not using DNSSEC or the other less common record types, you can
simplify the user interface even more by specifying a sequence of types the app
should use. Simply put this in your ``settings.py``::

  POWERDNS_RECORD_TYPES = (
      'A', 'AAAA', 'CNAME', 'HINFO', 'MX', 'NAPTR', 'NS',
      'PTR', 'SOA', 'SRV', 'TXT',
  )

Consult PowerDNS documentation for a `list of supported resource record types
<http://doc.powerdns.com/types.html>`_.

Notifications
------------------------

This app notifies owners of domains and records that were created for them. To
have them working you need to:

* enable them by setting ``ENABLE_OWNER_NOTIFICATIONS``

* configure ``FROM_EMAIL`` (the e-mail that would be a sender of the
  notifications) 

* configure ``OWNER_NOTIFICATIONS`` as following::

    OWNER_NOTIFICATIONS = {
        'Domain': (subject, content),
        'Record': (subject, content),
    }

* set up the `mailing backend
  <https://docs.djangoproject.com/en/1.8/topics/email/#email-backends>`_


In the subject and content strings you can place the following placeholders:

* ``object`` - the string representation of the domain or record created
* ``owner-email``
* ``owner-name``
* ``creator-email``
* ``creator-name``


Using a separate database for PowerDNS
--------------------------------------

If your Django application is using a different database than the one used by
PowerDNS, provide the configuration for the DNS database in ``settings.py`` as
a separate entry in ``DATABASES``, for example::

  DATABASES = {
      'default': {
          'ENGINE': 'django.db.backends.mysql',
          'NAME': 'project_db',
          'USER': 'user',
          'PASSWORD': 'secret',
          'HOST': '127.0.0.1',
          'PORT': '3306',
          'OPTIONS': {
              "init_command": "SET storage_engine=INNODB",
          },
      },
      'powerdns': {
          'ENGINE': 'django.db.backends.mysql',
          'NAME': 'powerdns',
          'USER': 'pdns',
          'PASSWORD': 'pdns',
          'HOST': '127.0.0.1',
          'PORT': '3306',
          'OPTIONS': {
              "init_command": "SET storage_engine=INNODB",
          },
      },
  }

For Django to automatically route ``powerdns`` requests to the right database,
add this setting to ``settings.py``::

  DATABASE_ROUTERS = ['powerdns.routers.PowerDNSRouter']

You have to sync and migrate the ``default`` and the ``powerdns`` databases
separately. First the default database::

  $ python manage.py syncdb
  $ python manage.py migrate

Then the ``powerdns`` database::

  $ python manage.py syncdb --database=powerdns
  $ python manage.py migrate --database==powerdns powerdns

Note that the ``powerdns`` database will maintain its own separate
migration history table. This is especially helpful if your connecting several
Django projects to a single PowerDNS database.



CHANGES.rst
Change Log
----------

0.9.3
~~~~~

* Fixed issue #3: HTTP 500 in record admin form if no type given

0.9.2
~~~~~

* Fixed issue #2: numeric sorting of IP addresses in admin

0.9.1
~~~~~

* Domain foreign keys support auto completion and have "Edit separately" links
  
* Field choices use radio selects whenever that makes sense (fever clicks
  necessary)

* Forward/reverse domain filter (requires Django 1.4+)

* The ``0002`` database migration in 0.9.0 was incomplete, this is now fixed

0.9.0
~~~~~

* DNSSEC tables supported.

* Support for multiple databases.

* Updated the project to require at least Django 1.3.

* UI translations supported (currently Polish translation added).

* South migrations.

* Source code compliant with PEP8.

* Minor fixes.

0.2.0
~~~~~

* First release with basic PowerDNS support.


dnsaas/settings.py
# Django settings for standalone django_powerdns.
import sys


DEBUG = True
TEMPLATE_DEBUG = DEBUG

TESTING = len(sys.argv) > 1 and sys.argv[1] == 'test'

ADMINS = (
    # ('Your Name', 'your_email@example.com'),
)

MANAGERS = ADMINS

if TESTING:
    DATABASES = {
        'default': {
            'ENGINE': 'django.db.backends.sqlite3',
            'NAME': ':memory:',
            'USER': '',
            'PASSWORD': '',
            'HOST': '',
            'PORT': '',
        }
    }
    EMAIL_BACKEND = 'django.core.mail.backends.locmem.EmailBackend'
else:
    DATABASES = {
        'default': {
            'ENGINE': 'django.db.backends.mysql',
            'NAME': 'dnsaas',
            'USER': 'dnsaas',
            'PASSWORD': 'dnsaas',
            'HOST': 'db',
            'PORT': '3306',
        }
    }

# Local time zone for this installation. Choices can be found here:
# http://en.wikipedia.org/wiki/List_of_tz_zones_by_name
# although not all choices may be available on all operating systems.
# On Unix systems, a value of None will cause Django to use the same
# timezone as the operating system.
# If running in a Windows environment this must be set to the same as your
# system time zone.
TIME_ZONE = 'America/Chicago'

# Language code for this installation. All choices can be found here:
# http://www.i18nguy.com/unicode/language-identifiers.html
LANGUAGE_CODE = 'en-us'

SITE_ID = 1

# If you set this to False, Django will make some optimizations so as not
# to load the internationalization machinery.
USE_I18N = True

# If you set this to False, Django will not format dates, numbers and
# calendars according to the current locale.
USE_L10N = True

# If you set this to False, Django will not use timezone-aware datetimes.
USE_TZ = False

# Absolute filesystem path to the directory that will hold user-uploaded files.
# Example: "/home/media/media.lawrence.com/media/"
MEDIA_ROOT = ''

# URL that handles the media served from MEDIA_ROOT. Make sure to use a
# trailing slash.
# Examples: "http://media.lawrence.com/media/", "http://example.com/media/"
MEDIA_URL = ''

# Absolute path to the directory static files should be collected to.
# Don't put anything in this directory yourself; store your static files
# in apps' "static/" subdirectories and in STATICFILES_DIRS.
# Example: "/home/media/media.lawrence.com/static/"
STATIC_ROOT = ''

# URL prefix for static files.
# Example: "http://media.lawrence.com/static/"
STATIC_URL = '/static/'

# Additional locations of static files
STATICFILES_DIRS = (
    # Put strings here, like "/home/html/static" or "C:/www/django/static".
    # Always use forward slashes, even on Windows.
    # Don't forget to use absolute paths, not relative paths.
)

# List of finder classes that know how to find static files in
# various locations.
STATICFILES_FINDERS = (
    'django.contrib.staticfiles.finders.FileSystemFinder',
    'django.contrib.staticfiles.finders.AppDirectoriesFinder',
)

# Make this unique, and don't share it with anybody.
SECRET_KEY = '5w%2cqg#kb1w&amp;mm-ss67(eo&amp;3+d9%pbu+5pesa*l&amp;pk7g-m48d'

# List of callables that know how to import templates from various sources.
TEMPLATE_LOADERS = (
    'django.template.loaders.filesystem.Loader',
    'django.template.loaders.app_directories.Loader',
)

MIDDLEWARE_CLASSES = (
    'django.middleware.common.CommonMiddleware',
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    # Uncomment the next line for simple clickjacking protection:
    # 'django.middleware.clickjacking.XFrameOptionsMiddleware',
)

ROOT_URLCONF = 'dnsaas.urls'

# Python dotted path to the WSGI application used by Django's runserver.
WSGI_APPLICATION = 'dnsaas.wsgi.application'

TEMPLATE_DIRS = ()

INSTALLED_APPS = (
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.sites',
    'django.contrib.messages',
    'django.contrib.staticfiles',
    'django_extensions',
    'django_nose',
    'rest_framework',
    'rest_framework.authtoken',
    'rest_framework_swagger',
    'powerdns',
    'dnsaas',
    'django.contrib.admin',
    'django.contrib.admindocs',
)

TEST_RUNNER = 'django_nose.NoseTestSuiteRunner'
NOSE_ARGS = [
    '-c.noserc',
    '--verbosity=2',
    '--exclude=integration-tests',  # To be run in docker
]

DNSAAS_DEFAULT_REVERSE_DOMAIN_TEMPLATE = 'reverse'

REST_FRAMEWORK = {
    'VIEW_DESCRIPTION_FUNCTION':
        'rest_framework_swagger.views.get_restructuredtext',
    'DEFAULT_AUTHENTICATION_CLASSES': (
        'rest_framework.authentication.BasicAuthentication',
        'rest_framework.authentication.TokenAuthentication',
    ),
    'DEFAULT_PERMISSION_CLASSES': (
        'rest_framework.permissions.IsAdminUser',
    ),
}


# A sample logging configuration. The only tangible logging
# performed by this configuration is to send an email to
# the site admins on every HTTP 500 error when DEBUG=False.
# See http://docs.djangoproject.com/en/dev/topics/logging for
# more details on how to customize your logging configuration.


import os
if os.environ.get('DJANGO_SETTINGS_PROFILE') == 'tests':
    DATABASES['default']['NAME'] = ':memory:'

ENABLE_OWNER_NOTIFICATIONS = TESTING  # E-mail backend required

FROM_EMAIL = 'dnsaas@example.com'

OWNER_NOTIFICATIONS = {
    'Domain': (
        'Domain {object} created for you!',
        """
        User {creator-name} ({creator-email}) has created a domain
        {object} and set its owner to you!
        """,
    ),
    'Record': (
        'Record {object} created for you!',
        """
        User {creator-name} ({creator-email}) has created a record
        {object} and set its owner to you!
        """,
    )
}


from settings_local import *  # noqa
